### PR TITLE
[Broker] Make PersistentTopicsBase#internalDeleteTopic async

### PR DESCRIPTION
**Pull request by RobertIndie**
Master Issue: #14013

### Motivation

See #14013

### Modifications

* Make PersistentTopicsBase#internalDeleteTopic async

### Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

### Documentation

- [x] `no-need-doc`

### CODE REVIEW

**Comment by shibd** [comment]
LGTM

**Review by liangyuanpeng — DISMISSED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 1]
```diff
@@ -311,20 +311,26 @@ protected void internalGrantPermissionsOnTopic(String role, Set<AuthAction> acti
...
+        validateTopicOwnershipAsync(topicName, authoritative)
+                .thenCompose(__ -> validateNamespaceOperationAsync(topicName.getNamespaceObject(),
+                        NamespaceOperation.DELETE_TOPIC))
+                .thenCompose(__ -> pulsar().getBrokerService().deleteTopic(topicName.toString(), true, deleteSchema))
+                .thenRun(() -> {
+                    log.info("[{}] Successfully removed topic {}", clientAppId(), topicName);
+                    asyncResponse.resume(Response.noContent().build());
+                })
+                .exceptionally(e -> {
+                    if (isManagedLedgerNotFoundException(e.getCause())) {
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:325]: Throwable throwable = FutureUtil.unwrapCompletionException(e);

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 2]
```diff
@@ -311,20 +311,26 @@ protected void internalGrantPermissionsOnTopic(String role, Set<AuthAction> acti
...
+                                                 boolean deleteSchema) {
+        validateTopicOwnershipAsync(topicName, authoritative)
+                .thenCompose(__ -> validateNamespaceOperationAsync(topicName.getNamespaceObject(),
+                        NamespaceOperation.DELETE_TOPIC))
+                .thenCompose(__ -> pulsar().getBrokerService().deleteTopic(topicName.toString(), true, deleteSchema))
+                .thenRun(() -> {
+                    log.info("[{}] Successfully removed topic {}", clientAppId(), topicName);
+                    asyncResponse.resume(Response.noContent().build());
+                })
+                .exceptionally(e -> {
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:324]: Throwable throwable = FutureUtil.unwrapCompletionException(e);

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 1 above)
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:325]:
```suggestion
                    if (isManagedLedgerNotFoundException(throwable)) {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 3]
```diff
@@ -311,20 +311,26 @@ protected void internalGrantPermissionsOnTopic(String role, Set<AuthAction> acti
...
+                    if (isManagedLedgerNotFoundException(e.getCause())) {
+                        log.info("[{}] Topic was already not existing {}", clientAppId(), topicName, e);
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:326]:
```suggestion
                        log.info("[{}] Topic was already not existing {}", clientAppId(), topicName, throwable);
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 4]
```diff
@@ -311,20 +311,26 @@ protected void internalGrantPermissionsOnTopic(String role, Set<AuthAction> acti
...
+                        asyncResponse.resume(Response.noContent().build());
+                    } else {
+                        log.error("[{}] Failed to delete topic forcefully {}", clientAppId(), topicName, e);
+                        asyncResponse.resume(new RestException(e));
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:330]:
```suggestion
                        resumeAsyncResponseExceptionally(asyncResponse, e);
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 5]
```diff
@@ -1007,32 +1013,36 @@ private void internalUnloadTransactionCoordinatorAsync(AsyncResponse asyncRespon
...
+    protected void internalDeleteTopic(AsyncResponse asyncResponse, boolean authoritative, boolean deleteSchema) {
+        validateNamespaceOperationAsync(topicName.getNamespaceObject(), NamespaceOperation.DELETE_TOPIC)
+                .thenCompose(__ -> validateTopicOwnershipAsync(topicName, authoritative))
+                .thenCompose(__ -> pulsar().getBrokerService().deleteTopic(topicName.toString(), false, deleteSchema))
+                .thenRun(() -> {
+                    log.info("[{}] Successfully removed topic {}", clientAppId(), topicName);
+                    asyncResponse.resume(Response.noContent().build());
+                })
+                .exceptionally(e -> {
+                    Throwable t = e.getCause();
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:1034]:
```suggestion
                    Throwable t = FutureUtil.unwrapCompletionException(e);
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 6]
```diff
@@ -1007,32 +1013,36 @@ private void internalUnloadTransactionCoordinatorAsync(AsyncResponse asyncRespon
...
+                    log.error("[{}] Failed to delete topic {}", clientAppId(), topicName, t);
+                    if (t instanceof TopicBusyException) {
+                        asyncResponse.resume(new RestException(Status.PRECONDITION_FAILED,
+                                "Topic has active producers/subscriptions"));
+                    } else if (isManagedLedgerNotFoundException(t)) {
+                        asyncResponse.resume(new RestException(Status.NOT_FOUND, "Topic not found"));
+                    } else {
+                        asyncResponse.resume(new RestException(t));
```
Comment (nodece) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:1042]:
```suggestion
                        resumeAsyncResponseExceptionally(asyncResponse, t);
```

**Comment by gaoran10** [comment]
Could we merge these two delete methods?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java [hunk 7]
```diff
@@ -1151,28 +1151,29 @@ public void testDeleteTopic() throws Exception {
...
         CompletableFuture<Void> deleteTopicFuture = new CompletableFuture<>();
         deleteTopicFuture.completeExceptionally(new MetadataStoreException.NotFoundException());
         doReturn(deleteTopicFuture).when(brokerService).deleteTopic(anyString(), anyBoolean(), anyBoolean());
-        persistentTopics.deleteTopic(testTenant, testNamespace, topicName, true, true, true);
-        //
+        AsyncResponse response = mock(AsyncResponse.class);
+        ArgumentCaptor<Response> responseCaptor = ArgumentCaptor.forClass(Response.class);
+        persistentTopics.deleteTopic(response, testTenant, testNamespace, topicName, true, true, true);
+        verify(response, timeout(5000).times(1)).resume(responseCaptor.capture());
+
```
Comment (nodece) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java:1162]: Please check the `responseCaptor.getValue().getStatus()`

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java [hunk 8]
```diff
@@ -1151,28 +1151,29 @@ public void testDeleteTopic() throws Exception {
...
-        try {
-            persistentTopics.deleteTopic(testTenant, testNamespace, topicName, true, true, true);
-        } catch (Exception e) {
-            Assert.assertTrue(e instanceof RestException);
-        }
-        //
+        response = mock(AsyncResponse.class);
+        persistentTopics.deleteTopic(response, testTenant, testNamespace, topicName, true, true, true);
+        verify(response, timeout(5000).times(1)).resume(errorCaptor.capture());
+
```
Comment (nodece) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java:1170]: Please check the `errorCaptor.getValue()`.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java [hunk 9]
```diff
@@ -1151,28 +1151,29 @@ public void testDeleteTopic() throws Exception {
...
         doReturn(deleteTopicFuture3).when(brokerService).deleteTopic(anyString(), anyBoolean(), anyBoolean());
-        try {
-            persistentTopics.deleteTopic(testTenant, testNamespace, topicName, false, true, true);
-        } catch (RestException e) {
-            Assert.assertEquals(e.getResponse().getStatus(), 404);
-        }
+        response = mock(AsyncResponse.class);
+        persistentTopics.deleteTopic(response, testTenant, testNamespace, topicName, false, true, true);
+        verify(response, timeout(5000).times(1)).resume(errorCaptor.capture());
     }
```
Comment (nodece) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java:1177]: Please check the `errorCaptor.getValue()`.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java [hunk 10]
```diff
@@ -1151,28 +1151,29 @@ public void testDeleteTopic() throws Exception {
...
         deleteTopicFuture3.completeExceptionally(new MetadataStoreException.NotFoundException());
         doReturn(deleteTopicFuture3).when(brokerService).deleteTopic(anyString(), anyBoolean(), anyBoolean());
-        try {
-            persistentTopics.deleteTopic(testTenant, testNamespace, topicName, false, true, true);
-        } catch (RestException e) {
-            Assert.assertEquals(e.getResponse().getStatus(), 404);
-        }
+        response = mock(AsyncResponse.class);
+        persistentTopics.deleteTopic(response, testTenant, testNamespace, topicName, false, true, true);
+        verify(response, timeout(5000).times(1)).resume(errorCaptor.capture());
```
Comment (nodece) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java:1176]: Please check the `errorCaptor.getValue()`.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java (hunk 8 above)
Comment (RobertIndie) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java:1173]: `errorCaptor` already checks that instance type.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/PersistentTopics.java [hunk 11]
```diff
@@ -301,13 +301,21 @@ public void unloadTopic(@Suspended final AsyncResponse asyncResponse, @PathParam
...
-            @QueryParam("force") @DefaultValue("false") boolean force,
-            @QueryParam("authoritative") @DefaultValue("false") boolean authoritative,
-            @QueryParam("deleteSchema") @DefaultValue("false") boolean deleteSchema) {
+    public void deleteTopic(@Suspended AsyncResponse asyncResponse, @PathParam("property") String property,
+                            @PathParam("cluster") String cluster,
+                            @PathParam("namespace") String namespace, @PathParam("topic") @Encoded String encodedTopic,
+                            @QueryParam("force") @DefaultValue("false") boolean force,
+                            @QueryParam("authoritative") @DefaultValue("false") boolean authoritative,
+                            @QueryParam("deleteSchema") @DefaultValue("false") boolean deleteSchema) {
         validateTopicName(property, cluster, namespace, encodedTopic);
```
Comment (liudezhi2098) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/PersistentTopics.java:310]: need catch exception

Comment (RobertIndie) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v1/PersistentTopics.java:310]: Same as https://github.com/apache/pulsar/pull/14064#discussion_r804353275

**Comment by RobertIndie** [comment]
These changes have been moved to https://github.com/apache/pulsar/pull/16232. So close this PR.

**Pull request closed without merging**